### PR TITLE
Added deployment of APIM traffic manager separated from control plane in pattern-4

### PR DESCRIPTION
**Pull request by hasuniea**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: advanced/am-pattern-4/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/Chart.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/requirements.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/requirements.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,82 @@
+{{/*
+Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/_helpers.tpl:2]:
```suggestion
Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-conf.yaml [hunk 4]
```diff
@@ -0,0 +1,171 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-conf.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,118 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-deployment.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-ingress.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-ingress.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-service.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/am/gateway/wso2am-pattern-4-am-gateway-service.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/am/control-plane/instance-2/wso2am-pattern-4-am-control-plane-deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  strategy:
+    type: Recreate
+  selector:
+    matchLabels:
+      deployment: {{ template "am-pattern-4.resource.prefix" . }}-am-cp
+      node: {{ template "am-pattern-4.resource.prefix" . }}-am-cp-2
+  template:
+    metadata:
+      annotations:
+        checksum.am.conf: {{ include (print $.Template.BasePath "/am/control-plane/instance-1/wso2am-pattern-4-am-control-plane-conf.yaml") . | sha256sum }}
```
Comment (ThilinaManamgoda) [inline advanced/am-pattern-4/templates/am/control-plane/instance-2/wso2am-pattern-4-am-control-plane-deployment.yaml:31]: Path should be `/am/control-plane/instance-2/wso2am-pattern-4-am-control-plane-conf.yaml`

**Review by chirangaalwis — CHANGES_REQUESTED** [review]

File: advanced/am-pattern-4/templates/mi/instance-1/wso2am-pattern-4-mi-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+              protocol: TCP
+          {{- end }}
+          {{- if .Values.wso2.deployment.mi.envs }}
+          env:
+            {{- range $key, $val := .Values.wso2.deployment.mi.envs }}
+            - name: {{ $key }}
+              value: {{ $val | quote }}
+          {{- end }}
+          {{- end }}
+      serviceAccountName: {{ .Values.kubernetes.serviceAccount }}
```
Comment (ThilinaManamgoda) [inline advanced/am-pattern-4/templates/mi/instance-1/wso2am-pattern-4-mi-deployment.yaml:102]: Do we need a service account for MI

File: advanced/am-pattern-4/values.yaml [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+           jvm:
+             # Resource allocation for the Java Heap
+             heap:
+                memory:
+                # Initial and minimum Heap size
+                  xms: "1024m"
+                  # Maximum Heap size
+                  xmx: "1024m"
+
+        # If the deployment configurations for the Gateway profile of WSO2 API Manager v3.2.0 (<WSO2AM>/repository/conf/deployment.toml),
```
Comment (ThilinaManamgoda) [inline advanced/am-pattern-4/values.yaml:181]: We are referring to APIM version `v3.2.0` here.

File: advanced/am-pattern-4/templates/mi/instance-1/wso2am-pattern-4-mi-deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,109 @@
+# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (ThilinaManamgoda) [inline advanced/am-pattern-4/templates/mi/instance-1/wso2am-pattern-4-mi-deployment.yaml:1]: Do we need a separate directory for for MI(`instance-1`) as we only have a single MI deployment?

File: advanced/am-pattern-4/templates/am/traffic-manager/instance-2/wso2am-pattern-4-am-trafficmanager-conf.yaml [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    auth_urls = ["ssl://{{ template "am-pattern-4.resource.prefix" . }}-am-cp-1-service:9711"]
+
+    [[apim.event_hub.publish.url_group]]
+    urls = ["tcp://{{ template "am-pattern-4.resource.prefix" . }}-am-cp-2-service:9611"]
+    auth_urls = ["ssl://{{ template "am-pattern-4.resource.prefix" . }}-am-cp-2-service:9711"]
+
+    [database.local]
+    url = "jdbc:h2:./repository/database/WSO2CARBON_DB;DB_CLOSE_ON_EXIT=FALSE"
+
+  {{- end }}
```
Comment (RakhithaRR) [inline advanced/am-pattern-4/templates/am/traffic-manager/instance-2/wso2am-pattern-4-am-trafficmanager-conf.yaml:165]: EOF missing

File: advanced/am-pattern-4/templates/am/control-plane/wso2am-pattern-4-am-control-plane-conf-entrypoint.yaml [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        if [[ -z "$(ls -A ${deployment_volume}/${shared_directory})" ]]; then
+          if ! cp -R ${original_deployment_artifacts}/${shared_directory}/* ${deployment_volume}/${shared_directory};
+          then
+            echo "Failed to copy the preserved, default artifacts to original location (${deployment_volume}/${shared_directory})"
+            exit 1
+          fi
+          echo "Successfully copied the preserved, default artifacts to original location (${deployment_volume}/${shared_directory})"
+        fi
+      fi
+    done
```
Comment (RakhithaRR) [inline advanced/am-pattern-4/templates/am/control-plane/wso2am-pattern-4-am-control-plane-conf-entrypoint.yaml:57]: We don't need those for 4.0.0

File: advanced/am-pattern-4/values.yaml [hunk 14]
```diff
@@ -0,0 +1,322 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/values.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/wso2am-pattern-4-service-account.yaml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/wso2am-pattern-4-service-account.yaml:1]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/wso2am-pattern-4-secrets.yaml [hunk 16]
```diff
@@ -0,0 +1,30 @@
+{{ if and (not (eq .Values.wso2.subscription.username "")) (not (eq .Values.wso2.subscription.password "")) }}
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/wso2am-pattern-4-secrets.yaml:2]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: advanced/am-pattern-4/templates/wso2am-pattern-4-mi-secrets.yaml [hunk 17]
```diff
@@ -0,0 +1,30 @@
+{{ if and (not (eq .Values.wso2.subscription.username "")) (not (eq .Values.wso2.subscription.password "")) (eq (default "docker.wso2.com" .Values.wso2.deployment.mi.dockerRegistry) "docker.wso2.com") }}
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/templates/wso2am-pattern-4-mi-secrets.yaml:2]:
```suggestion
# Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

**Review by chirangaalwis — CHANGES_REQUESTED** [review]

File: README.md [hunk 18]
```diff
@@ -15,10 +15,11 @@
 * [Deployment Pattern 1](advanced/am-pattern-1/README.md)
 * [Deployment Pattern 2](advanced/am-pattern-2/README.md)
 * [Deployment Pattern 3](advanced/am-pattern-3/README.md)
+* [Deployment Pattern 4](advanced/am-pattern-4/README.md)
 
 ## Changelog
 
-**Change log** from previous `v3.2.0.4` release: [View Here](https://github.com/wso2/kubernetes-apim/blob/3.2.x/CHANGELOG.md)
+**Change log** from previous `v3.2.0.4` release: [View Here](https://github.com/wso2/kubernetes-apim/blob/3.2.x/CHANGELOG.md)git
```
Comment (chirangaalwis) [inline README.md:22]:
```suggestion
**Change log** from previous `v4.0.0.1` release: [View Here](https://github.com/wso2/kubernetes-apim/blob/4.0.x/CHANGELOG.md)
```

File: advanced/am-pattern-4/Chart.yaml [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+appVersion: "4.0.0"
+description: A Helm chart for the deployment of WSO2 API Management deployment pattern 4
+name: am-pattern-4
+version: 4.0.0-1
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/Chart.yaml:19]:
```suggestion
version: 4.0.0-2
```

**Review by chirangaalwis — CHANGES_REQUESTED** [review]

File: README.md (hunk 18 above)
Comment (chirangaalwis) [inline README.md:22]: Since we are already adding this to the release note let's remove this entry for now.

```suggestion
```

File: README.md [hunk 20]
```diff
@@ -15,10 +15,11 @@
 * [Deployment Pattern 1](advanced/am-pattern-1/README.md)
 * [Deployment Pattern 2](advanced/am-pattern-2/README.md)
 * [Deployment Pattern 3](advanced/am-pattern-3/README.md)
+* [Deployment Pattern 4](advanced/am-pattern-4/README.md)
 
 ## Changelog
```
Comment (chirangaalwis) [inline README.md:20]:
```suggestion
```

File: README.md [hunk 21]
```diff
@@ -15,10 +15,11 @@
 * [Deployment Pattern 1](advanced/am-pattern-1/README.md)
 * [Deployment Pattern 2](advanced/am-pattern-2/README.md)
 * [Deployment Pattern 3](advanced/am-pattern-3/README.md)
+* [Deployment Pattern 4](advanced/am-pattern-4/README.md)
```
Comment (chirangaalwis) [inline README.md:19]:
```suggestion
```

**Review by chirangaalwis — CHANGES_REQUESTED** [review]

File: advanced/am-pattern-4/values.yaml [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+
+        # Kubernetes RollingUpdate strategy configurations
+        strategy:
+          rollingUpdate:
+            # The maximum number of pods that can be scheduled above the desired number of pods
+            maxSurge: 2
+            # The maximum number of pods that can be unavailable during the update
+            maxUnavailable: 0
+
+      # API Manager's Control Plane specific configurations
```
Comment (RakhithaRR) [inline advanced/am-pattern-4/values.yaml:144]: Should be traffic manager, not cp

File: advanced/am-pattern-4/values.yaml (hunk 10 above)
Comment (hasuniea) [reply advanced/am-pattern-4/values.yaml:181]: yes it should be for apim 3.2.0

File: advanced/am-pattern-4/values.yaml [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+        # Indicates whether the container is ready to service requests
+        readinessProbe:
+          # Number of seconds after the container has started before readiness probes are initiated
+          initialDelaySeconds: 180
+          # How often (in seconds) to perform the probe
+          periodSeconds: 10
+              # Number of deployment replicas
+        replicas: 2
+        resources:
+           # These are the minimum resource recommendations for running WSO2 API Management Control Plane deployment
```
Comment (RakhithaRR) [inline advanced/am-pattern-4/values.yaml:161]: Same here

File: advanced/am-pattern-4/requirements.yaml [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+dependencies:
+  - name: mysql-am
+    version: "4.0.0-1"
```
Comment (chirangaalwis) [inline advanced/am-pattern-4/requirements.yaml:17]:
```suggestion
    version: "4.0.0-2"
```

**Review by chirangaalwis — CHANGES_REQUESTED** [review]

**Pull request closed without merging**